docs/DesignDocs/PluggableAllocators.md

```diff
@@ -0,0 +1,142 @@
+- [The GNU Allocator](https://www.gnu.org/software/libc/manual/html_node/The-GNU-Allocator.html) which is based on `pthreads malloc`, which is in turn based on `dlmalloc`.
+- [Thread-Caching Malloc (tcmalloc)](https://gperftools.github.io/gperftools/tcmalloc.html) developed by Google.
+- [snmalloc](https://github.com/Microsoft/snmalloc) developed by Micrsoft Research.
+- and many more. Application specific allocators often outperform general-purpose allocators.
+
+
+This document describes the design of a scheme to allow custom allocators to be used in enclaves.
+
+
+Precedents
+------
+
+- The GNU C Library provides a way to replace its malloc implementation, documented in [Replacing Malloc](https://www.gnu.org/software/libc/manual/html_node/Replacing-malloc.html#Replacing-malloc). </br>
+  To replace the allocator, the following replacement functions must be provided `malloc`, `free`, `calloc`, `realloc`, `aligned_alloc`, `malloc_usable_size`, `memalign`, `posix_memalign`, `palloc`, and `valloc`. </br>
+  Replacement happens by statically linking the allocator library before the GNU C Library.
+
+- The CCF team has implemented compiling the Open Enclave SDK to use `snmalloc` (Pull Request #2649). Currently, users must recompile OE SDK with `USE_SNMALLOC` to leverage this.
+
+- Rust Language is [considering](https://doc.rust-lang.org/1.15.1/book/custom-allocators.html) capability to switch out the default allocators (`alloc_system` and `alloc_jemalloc`).
+
+
+Non-Requirements
+-----
+
+- This design does not aim to address what the ideal default allocator for Open Enclave should be.
+- Though the design could be extended to support multiple pluggable allocators, each managing a portion of the heap, it is currently a non-requirement.
+
+Requirements
+----
+
+The design for `Pluggable Allocators` additionally has the following requirements,
+
+- Allocators must be pluggable by an SDK user with having to recompile the OE SDK </br>
+  This is similat to the GNU C library in that, one need not rebuild glibc to plugin a malloc replacement.
+
+- Must work on various TEE environments that the Open Enclave SDK will support. </br>
+  Note, it does not mean that a given allocator must work on all platforms; rather Open Enclave SDK must support plugging allocators on all platforms. </br>
+  From a design perspective, this precludes the use of architecture specific datastructures in the plugin interface.
+
+- The design must support allocators to play well with features that Open Enclave SDK already provides (`debug allocators`, `backtracing`, `stack unwinding`, `failure callbacks` etc)
+
+- The design must allow allocators to perform thread-specific optimizations since multi-threaded applications benefit a lot from fine-tuned allocators.
+
+- Allocators must rely only on published interfaces and datastructures.
+
+- The design must allow for Open Enclave core (`liboecore`) to own a portion of the heap and have the allocator manage rest of the heap. </br>
```

Comment:
This requirement is achieved via `oe_allocator_init`. I have tried to explain how it works there.

The requirement for a part of the heap to be owned by `oecore` allows us to experiment with some ideas: stronger memory guarantess in oecore (never out of memory); multiple allocators;
allow supporting sandboxed enclaves etc.

I agree, we need to extended the API surface and harden this requirement once we support SGX 2.


I am unclear the impact SGX 2 will have on this design. I do suspect SGX 2 will impact a lot more than just memory allocators. It would be pragmatic to extend the interface
Do let me know if more clarification is needed. We can tweak the wording in subsequent PRs.

I agree that this design does not cover SGX V2; and we would need to expand the interface to support SGX V2. I don't think,